TiFlash crash in RandomFailPoints test dure tu memory tracker NPE

## Bug Report

Please answer these questions before submitting your issue. Thanks!
`
[2022/09/08 23:46:34.786 +08:00] [ERROR] [BaseDaemon.cpp:377 [BaseDaemon:########################################] [thread_id=2906]
[2022/09/08 23:46:34.786 +08:00] [ERROR] [BaseDaemon.cpp:378] ["BaseDaemon:(from thread 1727) Received signal Segmentation fault(11)."] [thread_id=2906]
[2022/09/08 23:46:34.786 +08:00] [ERROR] [BaseDaemon.cpp:406] ["BaseDaemon:Address: NULL pointer."] [thread_id=2906]
[2022/09/08 23:46:34.786 +08:00] [ERROR] [BaseDaemon.cpp:414] ["BaseDaemon:Access: read."] [thread_id=2906]
[2022/09/08 23:46:34.786 +08:00] [ERROR] [BaseDaemon.cpp:426] ["BaseDaemon:Unknown si_code."] [thread_id=2906]
[2022/09/08 23:46:34.947 +08:00] [ERROR] [BaseDaemon.cpp:570] ["BaseDaemon:
       0x653cce1faultSignalHandler(int, siginfo_t*, void*) [tiflash+106155233]
                libs/libdaemon/src/BaseDaemon.cpp:221
  0x7fb9488e7630<unknown symbol> [libpthread.so.0+63024]
       0x1cab9b6MemoryTracker::free(long) [tiflash+30063030]
                dbms/src/Common/MemoryTracker.cpp:163
       0x1c726edstd::__1::__shared_ptr_emplace<DB::TrackedMppDataPacket, std::__1::allocator<DB::TrackedMppDataPacket> >::__on_zero_shared() [tiflash+29828845]
                /usr/local/bin/../include/c++/v1/__memory/shared_ptr.h:315
       0x70a6eb1DB::AsyncTunnelSender::sendOne(bool) [tiflash+118124209]
                dbms/src/Flash/Mpp/MPPTunnel.cpp:402
       0x1d3cb2bDB::(anonymous namespace)::handleRpcs(grpc::ServerCompletionQueue*, std::__1::shared_ptr<DB::Logger> const&) [tiflash+30657323]
                dbms/src/Server/FlashGrpcServerHolder.cpp:54
       0x1d580bbDB::ExecutableTask<std::__1::packaged_task<void ()> >::execute() [tiflash+30769339]
                dbms/src/Common/ExecutableTask.h:52
       0x1d5b793DB::DynamicThreadPool::executeTask(std::__1::unique_ptr<DB::IExecutableTask, std::__1::default_delete<DB::IExecutableTask> >&) [tiflash+30783379]
                dbms/src/Common/DynamicThreadPool.cpp:101
       0x1d5adf0DB::DynamicThreadPool::fixedWork(unsigned long) [tiflash+30780912]
                dbms/src/Common/DynamicThreadPool.cpp:115
       0x1d5bee2void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, std::__1::thread DB::ThreadFactory::newThread<void (DB::DynamicThreadPool::*)(unsigned long), DB::DynamicThreadPool*, unsigned long&>(bool, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void (DB::DynamicThreadPool::*&&)(unsigned long), DB::DynamicThreadPool*&&, unsigned long&)::'lambda'(auto&&...), DB::DynamicThreadPool*, unsigned long> >(void*) [tiflash+30785250]
                /usr/local/bin/../include/c++/v1/thread:291
  0x7fb9488dfea5start_thread [libpthread.so.0+32421]"] [thread_id=2906]
`
### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiFlash version? (Required)
Git Commit Hash: 462b5450379a622a92982ad872c18680adbfbe30

--- a/dbms/src/Flash/Mpp/TrackedMppDataPacket.h
+++ b/dbms/src/Flash/Mpp/TrackedMppDataPacket.h
@@ -54,16 +54,14 @@ inline std::shared_ptr<MemoryTracker> getSharedPtrOfMemTracker(MemoryTracker * m
 struct MemTrackerWrapper
 {
     MemTrackerWrapper(size_t _size, MemoryTracker * memory_tracker)
-        : memory_tracker_holder(getSharedPtrOfMemTracker(memory_tracker))
-        , memory_tracker(memory_tracker)
+        : memory_tracker(getSharedPtrOfMemTracker(memory_tracker))
         , size(0)
     {
         alloc(_size);
     }
 
     explicit MemTrackerWrapper(MemoryTracker * memory_tracker)
-        : memory_tracker_holder(getSharedPtrOfMemTracker(memory_tracker))
-        , memory_tracker(memory_tracker)
+        : memory_tracker(getSharedPtrOfMemTracker(memory_tracker))
         , size(0)
     {}
 
@@ -95,8 +93,7 @@ struct MemTrackerWrapper
     {
         int bak_size = size;
         freeAll();
-        memory_tracker_holder = getSharedPtrOfMemTracker(new_memory_tracker);
-        memory_tracker = new_memory_tracker;
+        memory_tracker = getSharedPtrOfMemTracker(new_memory_tracker);
         alloc(bak_size);
     }
     ~MemTrackerWrapper()
@@ -109,8 +106,7 @@ struct MemTrackerWrapper
         free(size);
     }
 
-    std::shared_ptr<MemoryTracker> memory_tracker_holder;
-    MemoryTracker * memory_tracker;
+    std::shared_ptr<MemoryTracker> memory_tracker;
     size_t size = 0;
 };
 

--- a/dbms/src/Flash/Mpp/TrackedMppDataPacket.h
+++ b/dbms/src/Flash/Mpp/TrackedMppDataPacket.h
@@ -46,17 +46,24 @@ inline size_t estimateAllocatedSize(const mpp::MPPDataPacket & data)
     return ret;
 }
 
+inline std::shared_ptr<MemoryTracker> getSharedPtrOfMemTracker(MemoryTracker * memory_tracker)
+{
+    return memory_tracker ? memory_tracker->shared_from_this() : nullptr;
+}
+
 struct MemTrackerWrapper
 {
     MemTrackerWrapper(size_t _size, MemoryTracker * memory_tracker)
-        : memory_tracker(memory_tracker)
+        : memory_tracker_holder(getSharedPtrOfMemTracker(memory_tracker))
+        , memory_tracker(memory_tracker)
         , size(0)
     {
         alloc(_size);
     }
 
     explicit MemTrackerWrapper(MemoryTracker * memory_tracker)
-        : memory_tracker(memory_tracker)
+        : memory_tracker_holder(getSharedPtrOfMemTracker(memory_tracker))
+        , memory_tracker(memory_tracker)
         , size(0)
     {}
 
@@ -88,6 +95,7 @@ struct MemTrackerWrapper
     {
         int bak_size = size;
         freeAll();
+        memory_tracker_holder = getSharedPtrOfMemTracker(new_memory_tracker);
         memory_tracker = new_memory_tracker;
         alloc(bak_size);
     }
@@ -100,6 +108,8 @@ struct MemTrackerWrapper
     {
         free(size);
     }
+
+    std::shared_ptr<MemoryTracker> memory_tracker_holder;
     MemoryTracker * memory_tracker;
     size_t size = 0;
 };

